rust/connlib/tunnel/src/device_channel/tun_android.rs

```diff
@@ -100,16 +100,15 @@ impl ioctl::Request<GetInterfaceNamePayload> {
     }
 
     fn name(&self) -> std::borrow::Cow<'_, str> {
-        // Safety: The memory of `self.name` is always initialized.
-        let cstr = unsafe { std::ffi::CStr::from_ptr(self.name.as_ptr() as _) };
-
-        cstr.to_string_lossy()
```

Comment:
Very likely, there is something wrong with our unsafe code here. Given that there is a non-unsafe version, we should probably just do that.

Reply:
I guess what might be wrong is that the buffer is not always 0-terminated? In that case, we will read beyond array length and read invalid memory. See https://doc.rust-lang.org/std/ffi/struct.CStr.html#safety.

Reply:
That's what @conectado suspected too, but we tried all variants of the CStr functions with no luck, including the non-unsafe alternatives.

Strangely, the struct change here was the most reliable "fix" I found. I sent it for testing with a user so we can see what he reports.

There have been a few cases of compiler bugs for android in the past, so maybe it's that?

Reply:
See thread https://github.com/firezone/firezone/pull/3583#discussion_r1484198263

Reply:
> That's what @conectado suspected too, but we tried all variants of the CStr functions with no luck, including the non-unsafe alternatives.

In this patch, you are not using the `CStr` functions though but `String::from_utf8_lossy(&self.name)`. Are you saying that using `String::from_utf8_lossy(&self.name)` by itself does **not** fix the issue?

Reply:
Indeed, it does not. The problem is I believe that the memory that self.name references isn't valid anymore by the time the function returns, or gets "cleaned up" from the heap.

- does not happen in `debug`
- does not happen if we trace logs near the call
- does not happen on other targets just 32-bit android

Feels like a compiler bug being too eager to optimize heap utilization

Reply:
Let's pair today after standup on this if you're available

Reply:
User reported the crash is fixed with this build. It's a gnarly one for sure.